src/System.IO.FileSystem/src/System/IO/UnixFileStream.cs

```diff
@@ -403,19 +402,24 @@ private void VerifyOSHandlePosition()
         /// <param name="disposing">true to release both managed and unmanaged resources; false to release only unmanaged resources.</param>
         protected override void Dispose(bool disposing)
         {
-            // Flush and close the file
             try
             {
                 if (_fileHandle != null && !_fileHandle.IsClosed)
                 {
+                    // Flush any remaining data in the file
                     FlushWriteBuffer();
 
-                    // Unix doesn't directly support DeleteOnClose but we can mimick it.
-                    if ((_options & FileOptions.DeleteOnClose) != 0)
+                    // If DeleteOnClose was requested when constructed, delete the file now.
+                    // (Unix doesn't directly support DeleteOnClose, so we mimick it here.)
+                    if (_path != null && (_options & FileOptions.DeleteOnClose) != 0)
                     {
                         // Since we still have the file open, this will end up deleting
-                        // it (assuming we're the only link to it) once it's closed.
-                        Interop.Sys.Unlink(_path); // ignore any error
+                        // it (assuming we're the only link to it) once it's closed, but the
+                        // name will be removed immediatly.
+                        int result = Interop.Sys.Unlink(_path);
+                        Debug.Assert(result == 0,
+                            string.Format("unlink on {0} failed with result {1} and errno {2}",
+                                _path, result, Marshal.GetLastWin32Error()));
```

Comment:
Not sure, but Interop.Sys.GetLastError() wold give easy to read enum in error message, but could drop unknown error down to NONSTANDARD. Maybe add a ToString() to ErrorInfo and use that to print both the friendly converted enum name and the raw errno?